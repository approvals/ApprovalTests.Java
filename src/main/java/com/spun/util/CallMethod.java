package com.spun.util;

import java.util.ArrayList;

public class CallMethod
{
  private final Class<? extends Object>             clazz;
  private ArrayList<String>                         methodNames = new ArrayList<String>();
  private ArrayList<MethodExecutionPath.Parameters> params      = new ArrayList<MethodExecutionPath.Parameters>();
  public CallMethod(Class<? extends Object> clazz)
  {
    this.clazz = clazz;
  }
  public CallMethod method(String methodName, Object... parameters)
  {
    methodNames.add(methodName);
    if (!ArrayUtils.isEmpty(parameters))
    {
      params.add(new MethodExecutionPath.Parameters(parameters));
    }
    else {
      params.add(new MethodExecutionPath.Parameters((Object[])null));
    }
    return this;
  }
  public MethodExecutionPath go()
  {
    return new MethodExecutionPath(clazz, methodNames.toArray(new String[0]), params.toArray(new MethodExecutionPath.Parameters[0]));
  }
}
