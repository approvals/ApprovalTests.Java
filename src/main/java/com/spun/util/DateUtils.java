package com.spun.util;

import static org.mockito.Mockito.times;

import java.sql.Timestamp;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.TimeZone;

/**
 * A static class of convence functions for database access
 **/
public class DateUtils
{
  private static int TIME_SCALE[] = {Calendar.YEAR,
      Calendar.MONTH,
      Calendar.DATE,
      Calendar.HOUR_OF_DAY,
      Calendar.HOUR,
      Calendar.MINUTE,
      Calendar.SECOND,
      Calendar.MILLISECOND        };
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(Date firstDate, Date secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null)) { return (firstDate == secondDate); }
    return isSame(firstDate.getTime(), secondDate.getTime(), smallestUnits);
  }
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(Calendar firstDate, Calendar secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null)) { return (firstDate == secondDate); }
    return isSame(firstDate.getTime().getTime(), secondDate.getTime().getTime(), smallestUnits);
  }
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(Date firstDate, Calendar secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null)) { return ((Object) firstDate == (Object) secondDate); }
    return isSame(firstDate.getTime(), secondDate.getTime().getTime(), smallestUnits);
  }
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(Calendar firstDate, Date secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null)) { return ((Object) firstDate == (Object) secondDate); }
    return isSame(firstDate.getTime().getTime(), secondDate.getTime(), smallestUnits);
  }
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(long firstDate, long secondDate, int smallestUnits)
  {
    if (!ArrayUtils.contains(TIME_SCALE, smallestUnits)) { throw new Error("Invalid Timescale " + smallestUnits); }
    GregorianCalendar first = new GregorianCalendar();
    first.setTime(new Date(firstDate));
    setSignificantDigit(first, smallestUnits);
    GregorianCalendar second = new GregorianCalendar();
    second.setTime(new Date(secondDate));
    setSignificantDigit(second, smallestUnits);
    //     My_System.variable("Testing if " + first.getTime().getTime() + "==" + second.getTime().getTime());
    return (first.getTime().getTime() == second.getTime().getTime());
  }
  /************************************************************************/
  public static Timestamp getStartOfYear()
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    DateUtils.setSignificantDigit(gregorianCalendar, Calendar.YEAR);
    return new Timestamp(gregorianCalendar.getTime().getTime());
  }
  /************************************************************************/
  public static Timestamp getStartOfToday()
  {
    return getStartOfXDaysAgo(0);
  }
  /************************************************************************/
  public static Timestamp getStartOfXDaysAgo(int numberOfDays)
  {
    return getStartOfXDaysAgo(numberOfDays, new Date());
  }
  /************************************************************************/
  public static Timestamp getStartOfXDaysAgo(int numberOfDays, Date startingFrom)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(startingFrom);
    DateUtils.setSignificantDigit(gregorianCalendar, Calendar.DATE);
    gregorianCalendar.add(Calendar.DATE, -numberOfDays);
    return new Timestamp(gregorianCalendar.getTime().getTime());
  }
  /************************************************************************/
  public static Calendar getEndOfTodayAsCalendar()
  {
    return rollToEndOfDay(new Date());
  }
  /************************************************************************/
  public static Timestamp getEndOfToday()
  {
    return new Timestamp(getEndOfTodayAsCalendar().getTime().getTime());
  }
  /************************************************************************/
  public static Calendar setSignificantDigit(Date date, int smallestUnits)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTimeInMillis(date.getTime());
    return setSignificantDigit(calendar, smallestUnits);
  }
  /************************************************************************/
  public static Calendar setSignificantDigit(Calendar calendar, int smallestUnits)
  {
    boolean removeOn = false;
    for (int i = 0; i < TIME_SCALE.length; i++)
    {
      if (removeOn)
      {
        if (TIME_SCALE[i] == Calendar.DATE)
        {
          // Work arround since this is broke
          calendar.set(TIME_SCALE[i], 1);
        }
        else if (TIME_SCALE[i] == Calendar.HOUR_OF_DAY)
        {
          // Work arround since this is broke
          calendar.set(TIME_SCALE[i], 0);
        }
        else
        {
          calendar.clear(TIME_SCALE[i]);
        }
      }
      if (TIME_SCALE[i] == smallestUnits)
      {
        removeOn = true;
      }
    }
    return calendar;
  }
  /************************************************************************/
  public static Timestamp getStartOf(int unit, Date forDate)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTime(forDate);
    setSignificantDigit(calendar, unit);
    return new Timestamp(calendar.getTimeInMillis());
  }
  /************************************************************************/
  public static Timestamp getEndOf(int unit, Date forDate)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTime(forDate);
    setSignificantDigit(calendar, unit);
    calendar.add(unit, 1);
    calendar.add(Calendar.MILLISECOND, -1);
    return new Timestamp(calendar.getTimeInMillis());
  }
  /************************************************************************/
  public static boolean areSame(Date date1, Date date2, long accuracy)
  {
    return (compareDates(date1, date2) < accuracy);
  }
  /************************************************************************/
  /**
   * 
   * @return 1 if date1 > date2, 0 if date1 = date2, -1 if date1 < date2
   **/
  public static int compareDates(Date date1, Date date2)
  {
    long l1 = ((date1 == null) ? 0 : date1.getTime());
    long l2 = ((date2 == null) ? 0 : date2.getTime());
    long diff = l1 - l2;
    return (diff == 0) ? 0 : (int) (diff / Math.abs(diff));
  }
  /************************************************************************/
  public static Date createTime(Date date)
  {
    Calendar time = new GregorianCalendar();
    time.setTime(date);
    time.set(1970, 0, 1);
    return time.getTime();
  }
  /************************************************************************/
  /**
   * Rolls back till that time on a 24 hour clock
   **/
  public static Date rollTillHour(int hour, Date date)
  {
    Calendar rolled = new GregorianCalendar();
    rolled.setTime(date);
    setSignificantDigit(rolled, Calendar.HOUR_OF_DAY);
    while (rolled.get(Calendar.HOUR_OF_DAY) != hour)
    {
      rolled.add(Calendar.HOUR_OF_DAY, -1);
    }
    return rolled.getTime();
  }
  /************************************************************************/
  public static void main(String args[])
  {
    MySystem.variable("Calendar.DATE = " + Calendar.DATE);
    MySystem.variable("Year", setSignificantDigit(new GregorianCalendar(), Calendar.YEAR).getTime());
    MySystem.variable("Month", setSignificantDigit(new GregorianCalendar(), Calendar.MONTH).getTime());
    MySystem.variable("Day", setSignificantDigit(new GregorianCalendar(), Calendar.DAY_OF_MONTH).getTime());
    MySystem.variable("Hour", setSignificantDigit(new GregorianCalendar(), Calendar.HOUR).getTime());
    MySystem.variable("Minute", setSignificantDigit(new GregorianCalendar(), Calendar.MINUTE).getTime());
    MySystem.variable("End Of Day", rollToEndOfDay(new Date()).getTime());
  }
  /************************************************************************/
  public static GregorianCalendar rollToEndOfDay(Date date)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(date);
    gregorianCalendar.set(Calendar.HOUR_OF_DAY, 23);
    gregorianCalendar.set(Calendar.MINUTE, 59);
    gregorianCalendar.set(Calendar.SECOND, 59);
    gregorianCalendar.set(Calendar.MILLISECOND, 999);
    return gregorianCalendar;
  }
  /***********************************************************************/
  public static Timestamp asTimestamp(Date date)
  {
    return new Timestamp(date.getTime());
  }
  /************************************************************************/
  public static boolean isToday(Date date)
  {
    return DateUtils.isSame(date, new Date(), Calendar.DATE);
  }
  /************************************************************************/
  public static Calendar asCalendar(Date date)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(date);
    return gregorianCalendar;
  }
  /************************************************************************/
  public static Timestamp getLastOrCurrent(int dayOfWeek)
  {
    return getDayOfWeek(dayOfWeek, false);
    }
  /************************************************************************/
  public static Timestamp getNextOrCurrent(int dayOfWeek)
  {
    return getDayOfWeek(dayOfWeek, true);
  }
  /************************************************************************/
  private static Timestamp getDayOfWeek(int dayOfWeek, boolean foward) throws Error
  {
    int multiplier = foward  ? -1 : 1 ;
    for (int i = 0; i < 7; i++)
    {
      Timestamp day = getStartOfXDaysAgo(i * multiplier);
      if (asCalendar(day).get(Calendar.DAY_OF_WEEK) == dayOfWeek) { return day; }
    }
    throw new Error(String.format("didn't find a %s in the %s 7 days", dayOfWeek, foward ? "next":"last"));
  }
  /************************************************************************/
  /**
   * @param date "yyyy/MM/dd"
   */
  public static Timestamp parse(String date)
  {
    SimpleDateFormat format = new java.text.SimpleDateFormat("yyyy/MM/dd");
    try
    {
      return asTimestamp(format.parse(date));
    }
    catch (ParseException e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  /************************************************************************/
  /************************************************************************/
  
  public static boolean doesDaylightSavingsTimeStartOn(String date) {
    Timestamp day = parse(date);
    TimeZone timeZone = GregorianCalendar.getInstance().getTimeZone();
    boolean inDaylightTime = timeZone.inDaylightTime(day);
    Calendar tomorrow = asCalendar(day);
    tomorrow.add(Calendar.DATE, 1);
    Date time = tomorrow.getTime();
    return !inDaylightTime && timeZone.inDaylightTime(time);
  }
}