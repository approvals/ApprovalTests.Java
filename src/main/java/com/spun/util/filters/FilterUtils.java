package com.spun.util.filters;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

/**
 * Listens to the state of a EnabledConditions object
 **/
public class FilterUtils
{
  /***********************************************************************/
  /**
   * @return a new List containing all elements of the list for which isExtracted() would return true
   **/
  public static <T> ArrayList<T> retainExtracted(Collection<? extends T> fromList, Filter filter) throws IllegalArgumentException
  {
    return filter(fromList, filter, true);
  }
  /***********************************************************************/
  /**
   * @return a new List containing all elements of the list for which isExtracted() would return false
   **/
  public static <T> ArrayList<T> retainPurified(List<? extends T> fromList, Filter filter) throws IllegalArgumentException
  {
    return filter(fromList, filter, false);
  }
  /***********************************************************************/
  /**
   * @return a new List containing all elements of the list for which isExtracted() would return true
   **/
  public static <T> ArrayList<T> retainExtracted(T fromObjects[], Filter filter) throws IllegalArgumentException
  {
    return fromObjects == null ? new ArrayList<T>() : filter(Arrays.asList(fromObjects), filter, true);
  }
  /***********************************************************************/
  /**
   * @return a new List containing all elements of the list for which isExtracted() would return true
   **/
  private static <T> ArrayList<T> filter(Iterable<? extends T> collection, Filter filter, boolean retainExtracted) throws IllegalArgumentException
  {
    ArrayList<T> extracted = new ArrayList<T>();
    if (collection != null)
    {
      for (T t : collection)
      {
        boolean ex = (filter == null || filter.isExtracted(t));
        if (ex == retainExtracted)
        {
          extracted.add(t);
        }
      }
    }
    return extracted;
  }
  /***********************************************************************/
  /**
   * @return a new List containing all elements of the list for which isExtracted() would return false
   **/
  public static <T> ArrayList<T> retainPurified(T fromObjects[], Filter filter) throws IllegalArgumentException
  {
    return filter(Arrays.asList(fromObjects), filter, false);
  }
  /***********************************************************************/
  /*                         INNER CLASSES                               */
  /***********************************************************************/
  public static class FilterNotNull implements Filter
  {
    /***********************************************************************/
    public boolean isExtracted(Object object) throws IllegalArgumentException
    {
      return object != null;
    }
  }
  /***********************************************************************/
  /***********************************************************************/
}