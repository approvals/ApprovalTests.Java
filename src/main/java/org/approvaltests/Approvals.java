package org.approvaltests;

import java.awt.Component;
import java.awt.Image;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import org.apache.hadoop.fs.Path;
import org.approvaltests.ReporterFactory.FileTypes;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureOverrider;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.JUnitStackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.ApprovalBinaryFileWriter;
import org.approvaltests.writers.ApprovalTextWriter;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.approvaltests.writers.ComponentApprovalWriter;
import org.approvaltests.writers.DirectoryToDirectoryWriter;
import org.approvaltests.writers.FileApprovalWriter;
import org.approvaltests.writers.ImageApprovalWriter;
import org.approvaltests.writers.ResultSetApprovalWriter;
import org.jrack.RackResponse;
import org.jrack.RackResponseUtils;
import org.lambda.functions.Function1;
import org.lambda.functions.implementations.F1;
import org.lambda.query.Query;

import com.spun.util.ArrayUtils;
import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import com.spun.util.images.ImageWriter;
import com.spun.util.persistence.ExecutableQuery;
import com.spun.util.persistence.SqlLoader;

public class Approvals
{
  /**
   * @deprecated Use verify()
   */
  public static void approve(String response) throws Exception
  {
    verify(response);
  }
  public static void verify(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "txt"), FileTypes.Text);
  }
  public static void verify(Object o) throws Exception
  {
    verify("" + o);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String label, T[] array) throws Exception
  {
    verifyAll(label, array);
  }
  public static <T> void verifyAll(String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, String label, T[] array) throws Exception
  {
    verifyAll(header, label, array);
  }
  public static <T> void verifyAll(String header, String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(T[] values, Function1<T, String> f1)
  {
    verifyAll(values, f1);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, values, f1);
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, Arrays.asList(values), f1);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\r\n\r\n\r\n";
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String label, Iterable<T> array) throws Exception
  {
    verifyAll(label, array);
  }
  public static <T> void verifyAll(String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, String label, Iterable<T> array) throws Exception
  {
    verifyAll(header, label, array);
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Component c) throws Exception
  {
    verify(c);
  }
  public static void verify(Component c) throws Exception
  {
    verify(new ComponentApprovalWriter(c), FileTypes.Image);
  }
  /**
   * @deprecated Use verifyHtml()
   */
  public static void approveHtml(String response) throws Exception
  {
    verifyHtml(response);
  }
  public static void verifyHtml(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "html"), FileTypes.Html);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(File generateFile)
  {
    verify(generateFile);
  }
  public static void verify(File generateFile)
  {
    verify(new FileApprovalWriter(generateFile), FileTypes.File);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Image image)
  {
    verify(image);
  }
  public static void verify(Image image)
  {
    approve(ImageWriter.toBufferedImage(image), createApprovalNamer());
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(BufferedImage bufferedImage)
  {
    verify(bufferedImage);
  }
  public static void verify(BufferedImage bufferedImage)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(writter, namer, reporter);
  }
  public static void verify(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writter, namer), reporter);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ApprovalWriter writter, String fileType)
  {
    verify(writter, fileType);
  }
  public static void verify(ApprovalWriter writter, String fileType)
  {
    verify(writter, createApprovalNamer(), ReporterFactory.get(fileType));
  }
  /**
   * @deprecated Use verifyXml()
   */
  public static void approveXml(String xml) throws Exception
  {
    verifyXml(xml);
  }
  public static void verifyXml(String xml) throws Exception
  {
    verify(new ApprovalXmlWriter(xml), FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(FileApprover approver, ApprovalFailureReporter reporter)
  {
    verify(approver, reporter);
  }
  public static void verify(FileApprover approver, ApprovalFailureReporter reporter)
  {
    try
    {
      if (!approver.approve())
      {
        boolean passed = false;
        if (reporter instanceof ApprovalFailureOverrider)
        {
          passed = approver.askToChangeReceivedToApproved((ApprovalFailureOverrider) reporter);
        }
        if (!passed)
        {
          approver.reportFailure(reporter);
          approver.fail();
        }
        else
        {
          approver.cleanUpAfterSuccess(reporter);
        }
      }
      else
      {
        approver.cleanUpAfterSuccess(reporter);
      }
    }
    catch (Exception e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ExecutableQuery query) throws Exception
  {
    verify(query);
  }
  public static void verify(ExecutableQuery query) throws Exception
  {
    verify(new ApprovalTextWriter(query.getQuery(), "txt"), new JUnitStackTraceNamer(),
        new ExecutableQueryFailure(query));
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Map map) throws Exception
  {
    verify(map);
  }
  public static void verify(Map map) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(RackResponse response) throws Exception
  {
    verify(response);
  }
  public static void verify(RackResponse response) throws Exception
  {
    if (isImage(response))
    {
      String fileType = "png";
      verify(new ApprovalBinaryFileWriter(response.getResponse(), fileType), fileType);
    }
    else
    {
      verifyHtml(response.getResponse().toString());
    }
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ResultSet rs) throws Exception
  {
    verify(rs);
  }
  public static void verify(ResultSet rs) throws Exception
  {
    verify(new ResultSetApprovalWriter(rs), "csv");
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(SqlLoader loader) throws Exception
  {
    verify(loader);
  }
  public static void verify(SqlLoader loader) throws Exception
  {
    verify(new SqlLoader.ExecutableWrapper(loader));
  }
  private static boolean isImage(RackResponse response)
  {
    String type = response.getHeaders().get(RackResponseUtils.CONTENT_TYPE);
    return RackResponseUtils.CONTENT_TYPE_IMAGE.equals(type);
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return new JUnitStackTraceNamer();
  }
  private static void approve(BufferedImage bufferedImage, ApprovalNamer namer)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files) throws Error
  {
    ApprovalNamer namer = createApprovalNamer();
    String dirName = namer.getSourceFilePath() + Path.SEPARATOR + namer.getApprovalName() + ".Files";
    File approvedDirectory = new File(dirName);
    List<File> mismatched = new ArrayList<File>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(new DirectoryToDirectoryWriter(f, approvedDirectory), FileTypes.File);
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, new F1<File, String>(mismatched.get(1))
    {
      {
        ret(a.getName());
      }
    }).toString();
  }
}