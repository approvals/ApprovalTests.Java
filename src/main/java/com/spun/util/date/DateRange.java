package com.spun.util.date;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

import com.spun.util.DateUtils;
import com.spun.util.MethodExecutionPath;
import com.spun.util.ObjectUtils;
import com.spun.util.filters.Filter;
import com.spun.util.parser.TemplateDate;

public class DateRange
{
  public static final DateRange ALL = new DateRange(new Date(0), new Date(Long.MAX_VALUE));
  private final Date            end;
  private final Date            start;
  /************************************************************************/
  public DateRange(Date start, Date end)
  {
    this.start = start;
    this.end = end;
  }
  /************************************************************************/
  public Filter getFilter(Class clazz, String... methodCalls)
  {
    return getFilter(new MethodExecutionPath(clazz, methodCalls));
  }
  /************************************************************************/
  public Filter getFilter(MethodExecutionPath path)
  {
    return new DateRangeFilter(this, path);
  }
  /************************************************************************/
  public boolean contains(Date time)
  {
    time = (time == null) ? new Date() : time;
    return (start.getTime() <= time.getTime()) && (time.getTime() < end.getTime());
  }
  public static DateRange getRangeContaining(DateRange[] ranges, DateRange target)
  {
    for (DateRange dateRange : ranges)
    {
      if (dateRange.contains(target)) { return dateRange; }
    }
    return null;
  }
  private boolean contains(DateRange range)
  {
    return contains(range.end) || contains(range.start);
  }
  /************************************************************************/
  public Date getEnd()
  {
    return end;
  }
  /************************************************************************/
  public Date getStart()
  {
    return start;
  }
  /************************************************************************/
  @Override
  public int hashCode()
  {
    return ObjectUtils.generateHashCode(start.getTime(), end.getTime());
  }
  /************************************************************************/
  @Override
  public boolean equals(Object object)
  {
    if (this == object)
    {
      return true;
    }
    else if (object instanceof DateRange)
    {
      DateRange that = (DateRange) object;
      return this.start.getTime() == that.start.getTime() && this.end.getTime() == that.end.getTime();
    }
    else
    {
      return false;
    }
  }
  /************************************************************************/
  @Override
  public String toString()
  {
    return String.format("[%s - %s]", new TemplateDate(start).getDateAndTime(), new TemplateDate(end)
        .getDateAndTime());
  }
  /***********************************************************************/
  public DateRange[] getWeeks()
  {
    return getUnits(new WeekAware());
  }
  private DateRange[] getUnits(UnitAware unit)
  {
    ArrayList<DateRange> ranges = new ArrayList<DateRange>();
    Calendar start = DateUtils.asCalendar(getStart());
    Calendar end = (Calendar) start.clone();
    while (!(getEnd().getTime() <= end.getTime().getTime()))
    {
      end.add(Calendar.DAY_OF_YEAR, 1);
      if (unit.isStart(end) || getEnd().getTime() <= end.getTime().getTime())
      {
        ranges.add(new DateRange(start.getTime(), end.getTime()));
        start = (Calendar) end.clone();
      }
    }
    return ranges.toArray(new DateRange[ranges.size()]);
  }
  /***********************************************************************/
  public DateRange[] getMonths()
  {
    return getUnits(new MonthAware());
  }
  public DateRange[] getQuarters()
  {
    return getUnits(new QuarterAware());
  }
  /************************************************************************/
  public boolean containsDayOfWeek(int day)
  {
    return getFirst(day) != null;
  }
  public Date getFirst(int day)
  {
    Calendar start = DateUtils.asCalendar(getStart());
    while (start.getTime().getTime() <= getEnd().getTime())
    {
      start.add(Calendar.DAY_OF_YEAR, 1);
      if (start.get(Calendar.DAY_OF_WEEK) == day) { return start.getTime(); }
    }
    return null;
  }
  /************************************************************************/
  /*                            INNER CLASSES                             */
  /************************************************************************/
  public static class DateRangeFilter implements Filter
  {
    private final MethodExecutionPath path;
    private final DateRange           range;
    public DateRangeFilter(DateRange range, MethodExecutionPath path)
    {
      this.range = range;
      this.path = path;
    }
    public boolean isExtracted(Object object) throws IllegalArgumentException
    {
      return range.contains((Date) path.extractValue(object));
    }
  }
  public static interface UnitAware
  {
    public boolean isStart(Calendar end);
  }
  public static class WeekAware implements UnitAware
  {
    public boolean isStart(Calendar end)
    {
      return end.get(Calendar.DAY_OF_WEEK) == Calendar.SUNDAY;
    }
  }
  public static class MonthAware implements UnitAware
  {
    public boolean isStart(Calendar end)
    {
      return end.get(Calendar.DAY_OF_MONTH) == 1;
    }
  }
  public static class QuarterAware implements UnitAware
  {
    public boolean isStart(Calendar end)
    {
      return end.get(Calendar.DAY_OF_MONTH) == 1 && end.get(Calendar.MONTH) % 3 == 0;
    }
  }
  /************************************************************************/
  /************************************************************************/
}
