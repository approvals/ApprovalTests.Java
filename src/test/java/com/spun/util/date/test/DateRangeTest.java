package com.spun.util.date.test;

import java.sql.Timestamp;
import java.util.Calendar;
import java.util.Date;

import junit.framework.TestCase;

import org.approvaltests.Approvals;

import com.spun.util.DateUtils;
import com.spun.util.date.DateRange;
import com.spun.util.filters.FilterUtils;
import com.spun.util.tests.TestUtils;

public class DateRangeTest extends TestCase
{
  /************************************************************************/
  public void testIsIn() throws Exception
  {
    DateRange range = new DateRange(date(40), date(20));
    assertTrue(range.contains(date(40)));
    assertTrue(range.contains(date(30)));
    assertFalse(range.contains(date(20)));
    assertFalse(range.contains(date(10)));
  }
  /************************************************************************/
  public void testFilter() throws Exception
  {
    DateRange range = new DateRange(date(40), date(20));
    Timestamp dates[] = {date(50), date(40), date(30), date(20), date(10)};
    TestUtils.assertLength(2, FilterUtils.retainExtracted(dates, range.getFilter(Timestamp.class, "clone")));
  }
  /************************************************************************/
  public static Timestamp date(int daysAgo)
  {
    return DateUtils.getStartOfXDaysAgo(daysAgo);
  }
  /************************************************************************/
  public void testGetWeeks() throws Exception
  {
    DateRange d = new DateRange(DateUtils.parse("2008/10/01"), DateUtils.parse("2008/11/01"));
    Approvals.verifyAll("week", d.getWeeks());
  }
  /************************************************************************/
  public void testContainsDayOfWeek() throws Exception
  {
    DateRange d = new DateRange(DateUtils.parse("2008/11/11"), DateUtils.parse("2008/11/15"));
    assertTrue(d.containsDayOfWeek(Calendar.THURSDAY));
    assertFalse(d.containsDayOfWeek(Calendar.MONDAY));
  }
  /************************************************************************/
  public void testGetFirstDayOfWeek() throws Exception
  {
    DateRange d = new DateRange(DateUtils.parse("2008/11/11"), DateUtils.parse("2008/11/15"));
    Date expected = new Date(DateUtils.parse("2008/11/13").getTime());
    assertEquals(expected, d.getFirst(Calendar.THURSDAY));
  }
  /************************************************************************/
  public void testGetMonths() throws Exception
  {
    DateRange d = new DateRange(DateUtils.parse("2008/01/01"), DateUtils.parse("2009/01/01"));
    Approvals.verifyAll("months", d.getMonths());
  }
  /************************************************************************/
  public void testGetQuarters() throws Exception
  {
    DateRange d = new DateRange(DateUtils.parse("2008/01/01"), DateUtils.parse("2009/01/01"));
    Approvals.verifyAll("months", d.getQuarters());
  }
  /************************************************************************/
  public void testGetRangeContaining() throws Exception
  {
    DateRange d = new DateRange(DateUtils.parse("2008/01/01"), DateUtils.parse("2009/01/01"));
    DateRange containing = DateRange.getRangeContaining(d.getQuarters(), d.getMonths()[0]);
    Approvals.verify(containing.toString());
  }
  
  /************************************************************************/
  /************************************************************************/
}
