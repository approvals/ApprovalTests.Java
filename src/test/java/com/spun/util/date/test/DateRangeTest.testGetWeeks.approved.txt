week[0] = [10/01/2008 0:00 - 10/05/2008 0:00]
week[1] = [10/05/2008 0:00 - 10/12/2008 0:00]
week[2] = [10/12/2008 0:00 - 10/19/2008 0:00]
week[3] = [10/19/2008 0:00 - 10/26/2008 0:00]
week[4] = [10/26/2008 0:00 - 11/01/2008 0:00]
